### PR TITLE
Optimisation

### PR DESCRIPTION
**Pull request by jetuk**
This branch is getting a bit out of hand. 

I have addressed #16, #97, #101 and #103 here so far. There is also an MOEA example included.

Still to complete:
- [x] Add `Recorder` subclasses in MOEA example to `pywr.recorders`
- [x] Decide whether the `OptimisationModel` subclass should be left in the example or also added to `pywr` proper.

### CODE REVIEW

**Comment by jetuk** [comment]
Possibly the only outstanding thing is to add `Recorder.value` as a not implemented default method on the `Recorder` base class.

**Comment by snorfalorpagus** [comment]
The build failures are rounding errors. Should be easy to fix.

**Comment by jetuk** [comment]
Anything else to add to this now?

File: pywr/_core.pyx [hunk 1]
```diff
@@ -57,9 +57,11 @@ cdef class ScenarioCollection:
         def __get__(self):
             if len(self._scenarios) == 0:
                 return (1, )
-            return (sc._size for sc in self._scenarios)
+            return tuple(sc.size for sc in self._scenarios)
 
-    cpdef int ravel_indices(self, int[:] scenario_indices):
+    cpdef int ravel_indices(self, int[:] scenario_indices) except? -1:
```
Comment (snorfalorpagus) [inline pywr/_core.pyx]: Is `-1` a valid return from this function if an exception hasn't occured? If not the `?` isn't required.

**Comment by snorfalorpagus** [comment]
This looks good to me. Do you want to merge it?

File: pywr/_core.pyx (hunk 1 above)
Comment (jetuk) [inline pywr/_core.pyx]: You make a good point, but #109 rewrites this function anyway.

**Pull request merged**